README.md
# Generic CLI Toolset (In Python)

A simple but powerful Command Line Interface utility powered by argparse that will print the result to standard output (stdout), allowing scripting capabilities

- Usable in programs like dmenu or rofi

## Table of Contents
- [Information](#information)
- [Setup](#setup)
- [Documentation](#documentation)
- [Contacts](#contacts)

## Information

### Background

Powered by argparse, his toolset targets to be a generic CLI toolset as a foundation/base, allowing users to take and create their own functions that returns a variable. 
User just needs to modify the parser to include their personal cli arguments and they have made their own scriptable, portable, modular CLI program.

This project was inspired by my shellscripting adventures as well as seeing rofi/dmenu scripting capabilities.

I also aim to be able to use this toolset to create a simple lemonbar scripting environment for use in my workflow

### Features

+ Modifiable - this is a generic CLI toolset aiming to give users the convenience of making their own scriptable CLI 
+ portable - Just keep the base toolchain/toolset as a repository or backup and you can reuse it however much you like
+ scriptable - this is my main focus as I do use programs such as rofi and dmenu that allows for advance scripting capabilities as well as automation

### Program Information

+ Project Name : generic-cli-toolset

## Setup

### Pre-Requisites

### Dependencies
+ python
+ pip
+ argparse

### Installation/Compiling

## Documentation

### Using on its own

#### Synopsis/Syntax

python main.py [{options} [arguments]]

#### Parameters/Arguments

+ -h | --help : Displays this help message
+ -v | --version : Displays program version and other relevant information
+ -o | --get-opts : Returns all options in list format (stdout will be seperated by spaces)

#### Usage

+ Get options
	```console
	$> python main.py --get-opts
	$> list of options
	```

### Using as a Library

#### Import Library

- Import 'toolset'
	```python
	import toolset.toolset 
	```
- Import 'cliargs'
	```python
	import toolset.cliargs
	```

#### Functions

- toolset
	- Toolset
		- CLIArgControl
			+ generate_parser
			+ arg_add
			+ arg_parse

- cliargs
	- Initialize
		- get_parser()
		- get_args() 

### Customization/Configuration

- cliargs.Initialize
	```python
	arg_definitions = [
		## Arguments here

		# Argument 1
		{
			"short_opt" : "",
			"long_opt" : "",
			"optionals" : {
				# Your Optionals Here
			}
		},
		# Argument n
		{
			"short_opt" : "",
			"long_opt" : "",
			"optionals" : {
				# Your Optionals Here
			}
		}
	]
	```

## Contacts

- Author(s) : 
	- [Asura/Thanatisia](https://github.com/Thanatisia)

CHANGELOG.md
# CHANGELOG

## Table of Contents
- [Version History](#version-history)

## Version History
- 2022-04-25 1423H : v0.1.0
	- Initial Creation
	- Created README.md
	- Created Source File
	- Created Folder "src/lib"
- 2022-04-25 1555H : v0.2.0
	- Created library set [toolset]
	- Created library "cliargs" in lib/toolset : Wrapper and general Quality-of-Life-centric Helper library
	- Created library "toolset" in lib/toolset : General brain of the CLI Designer Toolset Utility
	- Source file showcases how to modify using just main.py, not necessary to modify the librar(ies) at all
- 2022-04-25 1718H : v0.3.0
    - Fixed indentation and standardization as well as consistency of functional dependencies.
    - Added parameter for generate_parser()



src/lib/toolset/cliargs.py
"""
CLIArgs Generic Command Line Interface (CLI) Toolset Library - Command Line Arguments Controller
"""

### Import Library/Modules ###

## Built-In Libraries
import sys
import os
import argparse

## Dependencies
import lib.toolset.toolset as toolset

class Initialize():
	"""
	Program Initialization

	1. Initialize argparse
	"""
	def __init__(self, parser=None, parser_opts=None, arguments=None, arg_list=None, namespace=None):
		"""
        :: Params
            parser : The parser you want to use (can be None)
                Type: argparse.ArgumentParser() object
                Optional: True
                Default: None

            parser_opts : Options to pass to the generating parser
                Type: Dictionary (Kwargs)
                Optional: True
                Default: None
                Values:
                    Key : Parameter Name
                    Value : Parameter value

            arguments : All Arguments and their definitions here
                Type: List
                Values:
                    [
                        # Argument 1
                        {
                            "flags" : ["flag-or-string"],
                            "optionals" : {
                                # ...
                            }
                        },
                        # Argument 2
                        {
                            "flags" : ["flag-or-string"],
                            "optionals" : {
                                # ...
                            }
                        },
                        ...
                    ]
                Optional: True
                Default: None

            arg_list : A List of strings to parse for parse_args()
                Type: List
                Optional: True
                Default: None

            namespace : A Namespace object to hold the variables
                Type: Namespace
                Optional: True
                Default: None
        """

        # Initialize Command Line Interface (CLI) Argument Controller class
        self.cliargs = toolset.Toolset.CLIArgsControl(parser, parser_opts)
        self.parser = self.cliargs.parser

        # Add Arguments
        number_of_arguments = len(arguments)
        for i in range(number_of_arguments):
            curr_arg = arguments[i]

            # Retrieve Argument Details
            # List
            curr_arg_name_or_flags = curr_arg["flags"]

            if "optionals" in curr_arg:
                # Dictionary
                curr_arg_optionals = curr_arg["optionals"]
            else:
                curr_arg_optionals = {}

            # Add current argument to Argument Parser
            self.cliargs.arg_add(curr_arg_name_or_flags, self.parser, **curr_arg_optionals)

        # Parse Arguments
        self.args = self.cliargs.arg_parse(self.parser, arg_list, namespace)

    def get_parser(self):
        return self.parser

    def get_args(self):
        return self.args


src/lib/toolset/toolset.py
"""
Library to the brains of the Generic Command Line Interface utility
"""

### Import Libraries/Modules ###

## Built-In
import sys
import os
import argparse

## Dependencies

class Toolset():
    """
    Generic CLI Toolset class
    """
    class CLIArgsControl():
        """
        CLI Argument Management wrapper of argparse
        """
        def __init__(self, parser=None, parser_opts=None):
            """
            Constructor

            :: Params
                parser : Personal parser to use without generating a new one 
                    Type: argparse.ArgumentParser()
                    Optional: True
                    Default: None

                parser_opts : Options to pass when generating the parser
                    Type: Dictionary
                    Optional: True
                    Default: None
                    Values:
                        Key : Parameters
                        Value : Argument
            """
            if parser == None:
                if parser_opts == None:
                    self.parser = self.generate_parser()
                else:
                    self.parser = self.generate_parser(**parser_opts)
            else:
                self.parser = parser

        def generate_parser(self, **parser_kwargs):
            """
            Initialize parser object
            """
            if parser_kwargs == None:
                parser = argparse.ArgumentParser()
            else:
                parser = argparse.ArgumentParser(**parser_kwargs)
            return parser

        def arg_add(self, name_or_flags=None, parser=None, **arg_defn):
            """
            :: Params
                #short_opt : Short Option name (i.e. -o, -h)
                #    Type: String
                # long_opt : Long Option name (i.e. --out, --help)
                #    Type: String
                name_or_flags : A Name/List of option strings (i.e. foo, -f, --foo)
                    Type: String/List
                parser : Your parser object
                    Type: argparse.ArgumentParser()
                    Default: self.parser
                arg_defn : Argument Definitions
                    Type: kwargs (Variable Length Dictionary argument)
            """
            if parser == None:
                # Default to using the constructor parser
                parser = self.parser
            parser.add_argument(name_or_flags, **arg_defn)
            return parser

        def arg_parse(self, parser, arg_list=None, namespace=None):
            """
            Parse the arguments parser

            :: Params
                parser: Your parser Object
                    Type: argparse.ArgumentParser()
                arg_list: List of strings to parse
                    Optional: Yes
                    Default: taken from sys.argv
                namespace: Namespace/object to take attributes (i.e. Namespace(variable="Value"))
                    Type: namespace object
                    Default: a new empty Namespace object
            """
            if arg_list == None and namespace == None:
                args = parser.parse_args()
            elif arg_list == None:
                args = parser.parse_args(namespace)
            elif namespace == None:
                args = parser.parse_args(arg_list)
            else:
                args = parser.parse_args(arg_list, namespace)
            return args


src/main.py
"""
A simple but powerful Command Line Interface utility powered by argparse that will print the result to standard output (stdout), allowing scripting capabilities

- Usable in programs like dmenu or rofi

:: Background

Powered by argparse, his toolset targets to be a generic CLI toolset as a foundation/base, allowing users to take and create their own functions that returns a variable. 
User just needs to modify the parser to include their personal cli arguments and they have made their own scriptable, portable, modular CLI program.

This project was inspired by my shellscripting adventures as well as seeing rofi/dmenu scripting capabilities.

I also aim to be able to use this toolset to create a simple lemonbar scripting environment for use in my workflow

:: Features
	+ Modifiable - this is a generic CLI toolset aiming to give users the convenience of making their own scriptable CLI 
	+ portable - Just keep the base toolchain/toolset as a repository or backup and you can reuse it however much you like
	+ scriptable - this is my main focus as I do use programs such as rofi and dmenu that allows for advance scripting capabilities as well as automation

:: Contacts
Author(s) : 
	- Asura (https://github.com/Thanatisia)

:: Dependencies
	+ python
	+ pip
	+ argparse

:: Program Design
	-h | --help : Displays this help message
	-v | --version : Prints the program's version and other program info
	-o | --get-opts : Returns all options in list format

:: Program Information
Program Name : generic-cli-toolset
Version History:
	- 2022-04-25 1434H : v0.1.0
		- Initial creation
"""

### Import Libraries/Modules ###

## Built-in Libaries
import sys
import os
import argparse

## Dependencies ##
import lib.toolset.toolset as toolset
import lib.toolset.cliargs as cliargs

options = {
	"help" : ["-h", "--help"],
	"version" : ["-v", "--version"],
}

def print_help():
    """
	Print the Help menu
	"""
    print("Help:")
    for label, opt_list in options.items():
        print("\t{} : {}".format(label, opt_list))

def print_version():
    """
    Print the version info
    """
    print("Version: {}, {} Made by {}".format("v0.1.0", "CLIArgs : Generic CLI Toolset (Name is horrible, sorry, still a WIP)", "Asura (Thanatisia)"))

def init():
    """
    Initialization
    """
    global genericli_toolset, genericli, cliargs_obj, args

    # Initialize classes
    genericli_toolset = toolset.Toolset()
    genericli = genericli_toolset.CLIArgsControl()

    # Initialize CLI Argument Controller
    cliargs_obj = cliargs.Initialize(None, 
        {
            "add_help" : False
        }, [
        {
            "flag" : ["-h", "--help"],
            "optionals" : {
                "metavar" : ''
            }
        },
        {
            "flag" : ["-v", "--version"],
            "optionals" : {
                "metavar" : ''
            }
        },	
        {
            "flag" : ["-o", "--get-opts"],
            "optionals" : {
                "metavar" : ''
            }
        },
    ])

    # Get arguments
    args = cliargs_obj.get_args()

    # Output

def setup():
    """
    Basic Setup
    """
    init()

def main():
    """ 
    Process Arguments
    """
    if args.help:
        # Display Help message
        print_help()
    elif args.version:
        # Display Version
        print_version()
    elif args.get_opts:
        # Output
        opt_str = ' '.join(options)
        print(opt_str)
    else:
        print_help()

if __name__ == "__main__":
    setup()
    main()
